src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,313 @@
+    t_upper += max<real_type>(0, floor_upper) * pi;
+
+    // Parameters that will be used in loop
+    int i = 1;
+
+    // Lambda used for calculating the roots using Regular Falsi Iteration
+    auto calc_t_intersect = [&](real_type t) {
+        real_type alpha = u * std::sin(t + a_) - v * std::cos(t + a_);
+        real_type beta = t * (u * std::cos(t + a_) + v * std::sin(t + a_));
+        real_type gamma = r_b_ * (alpha - beta);
+        return gamma + x * v - y * u;
+    };
+    IllinoisRootFinder find_root_between{calc_t_intersect, tol_conv};
+
+    // Iterate on roots
+    while (t_lower < tmax_)
+    {
+        // Find value in root function
+        real_type ft_lower = calc_t_intersect(t_lower);
+        real_type ft_upper = calc_t_intersect(t_upper);
+
+        // Only iterate when roots have different signs
+        if (signum<real_type>(ft_lower) != signum<real_type>(ft_upper))
+        {
+            // Regular Falsi Iteration: Sometimes will slowly converge
+            real_type t_gamma = find_root_between(t_lower, t_upper);
+
+            // Convert root to distance and store if positive and in interval
+            real_type dist = calc_dist(x, y, u, v, t_gamma);
+            if (dist > tol_point
+                || (!(on_surface == SurfaceState::on) && (dist > 0)))
+            {
+                result[j] = convert * dist;
+                j++;
+            }
+
+            // Obtain next bounds
+            t_lower = t_upper;
+            t_upper += pi;
+        }
+        else
+        {
+            // Incremet interval slowly until root is in interval
+            // i used to slow down approach to next root
+            t_lower = t_upper;
+            t_upper += pi / i;
+            i++;
+        }
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the line-angle parameter \em beta used to find bounds of roots.
+ *
+ * \f[ \beta = \arctan(-v/u) \f]
+ */
+CELER_FUNCTION real_type InvoluteSolver::line_angle_param(real_type u,
+                                                          real_type v)
+{
+    using constants::pi;
+
+    if (u != 0)
+    {
+        // Standard method
+        return std::atan(-v / u);
+    }
+    else if (-v < 0)
+    {
+        // Edge case
+        return pi * -0.5;
+    }
+    else
+    {
+        return pi * 0.5;
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert root to distance by calculating the point on the involute given by
+ * the root and then taking the distance to that point from the particle.
+ */
+CELER_FUNCTION real_type InvoluteSolver::calc_dist(
+    real_type x, real_type y, real_type u, real_type v, real_type t) const
+{
+    real_type angle = t + a_;
+    real_type x_inv = r_b_ * (std::cos(angle) + t * std::sin(angle));
+    real_type y_inv = r_b_ * (std::sin(angle) - t * std::cos(angle));
```

Comment:
This function evaluation is repeated in at least three different places between here and the involute solver. Maybe a little helper class would be useful:
```c++
class NormalizedInvolute
{
inline CELER_FUNCTION Real2 calc_point(real_type theta) const;

real_type r_b_;
real_type a_;
};
```